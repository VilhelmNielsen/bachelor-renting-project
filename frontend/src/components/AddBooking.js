/** @jsx jsx */
import { jsx, css } from '@emotion/core'
import { useState } from 'react'
import { useMutation } from 'react-apollo-hooks'
import graphql from 'graphql-tag'
import useInput from '../lib/useInput'
import Error from './Error'
import Button from './elements/Button'
import { Form, Label, Input, Fieldset } from './elements/Form'

const MUTATION_BOOK_ITEM = graphql`
  mutation bookItem($id: ID!, $startDate: DateTime!, $endDate: DateTime!) {
    book(itemId: $id, startDate: $startDate, endDate: $endDate) {
      id
    }
  }
`

// TODO: Add calendar for start date
// TODO: Add calendar for end date
function AddBooking(props) {
  const [startDate, onChangeStartDate] = useInput('')
  const [endDate, onChangeEndDate] = useInput('')
  const [busy, setBusy] = useState(false)
  const [error, setError] = useState(null)

  const bookItem = useMutation(MUTATION_BOOK_ITEM, {
    variables: {
      id: props.id,
      startDate,
      endDate,
    },
  })

  const styles = {
    container: css`
      padding: 1rem;
      border: 1px solid hsl(0, 0%, 80%);
    `,
    title: css``,
  }

  return (
    <div css={styles.container} {...props}>
      <h3 css={styles.title}>Book this!</h3>
      {error && <Error error={error} />}
      <Form
        onSubmit={async event => {
          event.preventDefault()
          setBusy(true)
          setError(null)
          try {
            await bookItem()
          } catch (error) {
            setError(error)
          } finally {
            setBusy(false)
          }
        }}
      >
        <Fieldset disabled={busy}>
          <Label htmlFor="startDate">
            Start Date
            <Input
              required
              type="date"
              id="startDate"
              name="startDate"
              value={startDate}
              onChange={onChangeStartDate}
            />
          </Label>
          <Label htmlFor="endDate">
            End Date
            <Input
              required
              type="date"
              id="endDate"
              name="endDate"
              value={endDate}
              onChange={onChangeEndDate}
            />
          </Label>
          <Button>Book{busy && 'ing'} it!</Button>
        </Fieldset>
      </Form>
    </div>
  )
}

export default AddBooking
