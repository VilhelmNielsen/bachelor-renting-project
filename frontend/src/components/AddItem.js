import React, { useState } from 'react'
import gql from 'graphql-tag'
import { Input, Label, Textarea, Form } from './elements/Form'
import Button from './elements/Button'
import { useMutation } from 'react-apollo-hooks'

const MUTATION_ADD_ITEM = gql`
  mutation addItem(
    $title: String!
    $description: String!
    $price: Int!
    $maxDuration: Int
    $category: ID!
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      maxDuration: $maxDuration
      category: $category
    ) {
      id
    }
  }
`

function useInput(initialValue) {
  const [value, setValue] = useState(initialValue)
  const onChange = event => setValue(event.target.value)
  return [value, onChange, setValue]
}

function AddItem() {
  const [durationToggle, setDurationToggle] = useState(false)
  const [title, onChangeTitle, setTitle] = useInput('')
  const [description, onChangeDescription, setDescription] = useInput('')
  const [price, onChangePrice, setPrice] = useInput(0)
  const [maxDuration, onChangeMaxDuration, setMaxDuration] = useInput(0)
  const [busy, setBusy] = useState(false)
  // TODO: Make category dynamic
  const category = 'cjpi9rl3au3tq0a57j68jfr8q'
  const variables = { title, description, category, price: Number(price) }

  const addItem = useMutation(MUTATION_ADD_ITEM, {
    variables: durationToggle
      ? { ...variables, maxDuration: Number(maxDuration) }
      : variables,
  })
  return (
    <div>
      <Form
        onSubmit={async event => {
          event.preventDefault()
          try {
            setBusy(true)
            const res = await addItem()
            setTitle('')
            setDescription('')
            setPrice(0)
            setMaxDuration(0)
            // TODO: Success!
          } catch (error) {
            // TODO: Handle errors
            console.error(error.message)
          } finally {
            setBusy(false)
          }
        }}
      >
        <Label htmlFor="title">
          Title
          <Input
            autoComplete="off"
            value={title}
            onChange={onChangeTitle}
            type="text"
            name="title"
            id="title"
          />
        </Label>

        <Label htmlFor="description">
          Description
          <Textarea
            autoComplete="off"
            value={description}
            onChange={onChangeDescription}
            id="description"
          />
        </Label>
        <Label htmlFor="price">
          Price
          <Input
            autoComplete="off"
            value={price}
            onChange={onChangePrice}
            type="number"
            name="price"
            id="price"
          />
        </Label>
        <Label htmlFor="durationToggle">
          Max duration?
          <Input
            autoComplete="off"
            checked={durationToggle}
            onChange={event => setDurationToggle(event.target.checked)}
            type="checkbox"
            name="durationToggle"
            id="durationToggle"
          />
        </Label>
        {durationToggle && (
          <Label htmlFor="maxDuration">
            Max Duration in days
            <Input
              autoComplete="off"
              value={maxDuration}
              onChange={onChangeMaxDuration}
              type="number"
              name="maxDuration"
              id="maxDuration"
            />
          </Label>
        )}
        <Button disabled={busy}>Add Item</Button>
      </Form>
    </div>
  )
}

export default AddItem
