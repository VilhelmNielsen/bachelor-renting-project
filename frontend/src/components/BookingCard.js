/** @jsx jsx */
import { Suspense, useState } from 'react'
import { jsx, css } from '@emotion/core'
import { Img } from 'the-platform'
import { differenceInCalendarDays, format } from 'date-fns'
import formatPrice from '../lib/formatPrice'
import ReviewItem from './ReviewItem'
import { UnstyledButton } from './elements/Button'

const styles = {
  card: css`
    position: relative;
    background-color: white;
    box-shadow: 2px 2px 35px hsla(0, 0%, 0%, 0.05),
      2px 2px 20px hsla(0, 0%, 0%, 0.1);
    z-index: 1;
    transition: all 300ms;
  `,
  cardHovered: css`
    box-shadow: 2px 2px 60px hsla(0, 0%, 0%, 0.07),
      2px 2px 40px hsla(0, 0%, 0%, 0.15);
    transform: translateY(-1px);
  `,
  link: css`
    color: inherit;
    text-decoration: inherit;
  `,
  title: css`
    margin-top: 0;
    margin-bottom: 0;
  `,
  rating: css`
    margin: 0;
  `,
  image: css`
    background-color: #f8f9f8;
    height: 100%;
    width: 100%;
    object-fit: cover;
  `,
  imageContainer: css`
    position: relative;
    height: 250px;
  `,
  contentContainer: css`
    padding: 0.5rem 1rem;
  `,
  status: css`
    position: absolute;
    bottom: 0;
    right: 0;
    padding: 0 1rem;
    font-size: 0.8rem;
    background: #898989;
    color: #ffffff;
    line-height: 2;
    border-radius: 6px 0 0 0;
  `,
  statusGreen: css`
    background: #3d9951;
    color: #fcfff5;
  `,
  statusRed: css`
    background: #993d3d;
    color: #fff5f9;
  `,
  reviewContainer: css`
    position: absolute;
    width: 100%;
    left: 0;
    top: 100%;
    padding: 1rem;
    background-color: white;
    box-shadow: 2px 2px 35px hsla(0, 0%, 0%, 0.05),
      2px 2px 20px hsla(0, 0%, 0%, 0.1);
  `,
  price: css`
    font-size: 1.5em;
    font-weight: 600;
  `,
}

function Status({ type, ...props }) {
  switch (type) {
    case 'APPROVED':
      return <div css={[styles.status, styles.statusGreen]}>Approved</div>
    case 'CANCELLED':
      return <div css={[styles.status, styles.statusRed]}>Cancelled</div>
    case 'DENIED':
      return <div css={[styles.status, styles.statusRed]}>Denied</div>
    default:
      return <div css={styles.status}>Pending</div>
  }
}

function BookingCard(props) {
  const [hovered, setHovered] = useState(false)
  const [review, setReview] = useState(false)
  const { item, payment, status, startDate, endDate } = props.booking
  const duration = differenceInCalendarDays(endDate, startDate)
  return (
    <div
      onMouseEnter={() => setHovered(true)}
      onMouseLeave={() => setHovered(false)}
      css={styles.card}
    >
      <div css={styles.imageContainer}>
        {item.image ? (
          <Suspense
            maxDuration={100}
            fallback={
              <img src={item.image.preview} alt="" css={styles.image} />
            }
          >
            <Img src={item.image.full} alt="" css={styles.image} />
          </Suspense>
        ) : (
          <div css={styles.image}>No image...</div>
        )}
        <Status type={status} />
      </div>
      <div css={styles.contentContainer}>
        <h2 css={styles.title}>{item.title}</h2>
        <p>
          Renting for <strong>{duration}</strong> day
          {duration > 1 && 's'}
        </p>
        <p
          css={css`
            text-indent: 2rem;
          `}
        >
          <time>{format(startDate, 'DD/MM/YY')}</time> â€“{' '}
          <time>{format(endDate, 'DD/MM/YY')}</time>
        </p>
        <p>
          For <span css={styles.price}>{formatPrice(payment.price)}</span>
        </p>
        <div
          css={css`
            border-bottom: 1px solid #eeeeee;
            margin: 1rem 0 0.5rem;
          `}
        />
        <UnstyledButton
          style={{ marginBottom: '1rem', width: '100%' }}
          onClick={() => setReview(review => !review)}
        >
          Review
        </UnstyledButton>
        {review && (
          <div css={styles.reviewContainer}>
            <ReviewItem id={item.id} />
          </div>
        )}
      </div>
    </div>
  )
}

export default BookingCard
