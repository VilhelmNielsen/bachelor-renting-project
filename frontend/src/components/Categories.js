/** @jsx jsx */
import { useEffect } from 'react'
import { jsx, css } from '@emotion/core'
import { Link } from '@reach/router'
import gql from 'graphql-tag'
import { useQuery } from 'react-apollo-hooks'
import { siteMeta } from '../config'

import Layout from './Layout'

function Categories() {
  const {
    data: { categories },
    refetch,
  } = useQuery(QUERY_CATEGORIES, { suspend: true })

  useEffect(() => {
    document.title = 'Categories'
    return () => (document.title = siteMeta.title)
  }, [])

  useEffect(() => {
    refetch()
  }, [])

  return (
    <Layout>
      <section>
        <h1>Categories</h1>
        {categories &&
          categories.map(category => (
            <CategoryCard category={category} key={category.slug} />
          ))}
      </section>
    </Layout>
  )
}

const QUERY_CATEGORIES = gql`
  query categoriesPage {
    categories {
      id
      slug
      title
    }
  }
`

const cardStyles = {
  card: css`
    padding: 1rem;
  `,
}

function CategoryCard(props) {
  const { category } = props
  return (
    <div css={cardStyles.card}>
      <Link to={`/category/${category.slug}`}>{category.title}</Link>
    </div>
  )
}

export default Categories
export { QUERY_CATEGORIES }
