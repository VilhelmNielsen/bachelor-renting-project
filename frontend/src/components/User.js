import React, { createContext, useContext } from 'react'
import { useQuery, useMutation } from 'react-apollo-hooks'
import graphql from 'graphql-tag'

const MUTATION_SIGNIN = graphql`
  mutation signIn($email: String!, $password: String!) {
    signIn(email: $email, password: $password) {
      id
      name
    }
  }
`

const MUTATION_SIGNOUT = graphql`
  mutation signOut {
    signOut {
      message
    }
  }
`

const MUTATION_SIGNUP = graphql`
  mutation signUp(
    $name: String!
    $email: String!
    $password: String!
    $imageFull: String
    $imagePreview: String
  ) {
    signUp(
      name: $name
      email: $email
      password: $password
      imageFull: $imageFull
      imagePreview: $imagePreview
    ) {
      id
      name
    }
  }
`

const QUERY_CURRENT_USER = graphql`
  query currentUser {
    me {
      id
      name
    }
  }
`

export const userContext = createContext({ user: undefined })

export const useUser = () => useContext(userContext)

const User = {}

function UserProvider(props) {
  const {
    data: { me },
  } = useQuery(QUERY_CURRENT_USER, { errorPolicy: 'ignore' })
  const signUpMutation = useMutation(MUTATION_SIGNUP)
  const signInMutation = useMutation(MUTATION_SIGNIN)
  const signOutMutation = useMutation(MUTATION_SIGNOUT)

  const signIn = async ({ email, password }) => {
    await signInMutation({
      variables: { email, password },
      refetchQueries: [{ query: QUERY_CURRENT_USER }],
    })
  }

  const signOut = async () => {
    await signOutMutation({
      refetchQueries: [{ query: QUERY_CURRENT_USER }],
    })
  }

  const signUp = async ({ email, name, password, imageFull, imagePreview }) => {
    await signUpMutation({
      variables: { email, password, name, imageFull, imagePreview },
      refetchQueries: [{ query: QUERY_CURRENT_USER }],
    })
  }

  return (
    <userContext.Provider
      {...props}
      value={{ user: me, signIn, signOut, signUp }}
    />
  )
}

User.Provider = UserProvider
User.Consumer = userContext.Consumer

export default User
export {
  QUERY_CURRENT_USER,
  MUTATION_SIGNUP,
  MUTATION_SIGNIN,
  MUTATION_SIGNOUT,
}
