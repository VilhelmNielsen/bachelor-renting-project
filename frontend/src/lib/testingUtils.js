import casual from 'casual'
import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { MockLink } from 'apollo-link-mock'

casual.seed(666)

export const fakeUser = overrides => ({
  id: 'abc123',
  name: casual.full_name,
  email: casual.email,
  items: [],
  bookings: [],
  pendingBookings: [],
  ...overrides,
})

export const fakeCategory = overrides => ({
  id: 'cat345',
  title: casual.title,
  slug: casual.word,
  description: casual.description,
  items: [],
  ...overrides,
})

export function createClient(mocks) {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new MockLink(mocks),
  })
}
