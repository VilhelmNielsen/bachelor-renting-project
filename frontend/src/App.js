/** @jsx jsx */
import { jsx, css } from '@emotion/core'
import { lazy, Suspense, Fragment } from 'react'
import { Router } from '@reach/router'
import Spinner from './components/Spinner'
import Header from './components/Header'
import User from './components/User'

const Item = lazy(() => import('./components/Item'))
const Categories = lazy(() => import('./components/Categories'))
const Category = lazy(() => import('./components/Category'))
const AddItem = lazy(() => import('./components/AddItem'))
const SignIn = lazy(() => import('./components/SignIn'))
// import Search from './components/Search'

function App() {
  return (
    <Suspense fallback={<Spinner />}>
      <User.Provider>
        <Fragment>
          <Header />
          <h1
            css={css`
              color: hotpink;
            `}
          >
            Hello
          </h1>
          <Suspense fallback={<Spinner />}>
            <Router>
              <Categories path="/" />
              <AddItem path="add-item" />
              <Item path="item/:id" />
              <Category path="category/:slug" />
              <SignIn path="sign-in" />
            </Router>
          </Suspense>
        </Fragment>
      </User.Provider>
    </Suspense>
  )
}

export default App
