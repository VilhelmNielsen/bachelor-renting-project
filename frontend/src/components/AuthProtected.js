import React, { lazy } from 'react'
import { useUser } from './User'
import Layout from './Layout'

const SignIn = lazy(() => import('./SignIn'))

function AuthProtected(props) {
  const { user } = useUser()
  if (!user) {
    return (
      <Layout>
        <h1>You need to sign in to see this</h1>
        <SignIn />
      </Layout>
    )
  }

  return props.children
}

export default AuthProtected
