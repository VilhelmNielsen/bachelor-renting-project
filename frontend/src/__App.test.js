import React from 'react'
import { ApolloProvider } from 'react-apollo-hooks'
import { render, wait } from 'react-testing-library'
import { createClient, fakeCategory } from './lib/testingUtils'
import App from './App'
import { QUERY_CURRENT_USER } from './components/User'
import { QUERY_CATEGORIES } from './components/Categories'

const categoryMock = {
  request: { query: QUERY_CATEGORIES },
  result: { data: { categories: [fakeCategory()] } },
}

const unauthenticatedMock = [
  {
    request: {
      query: QUERY_CURRENT_USER,
    },
    result: {
      data: { me: null },
    },
  },
  categoryMock,
]

describe('<App />', () => {
  it('renders and matches snapshot', async () => {
    const client = createClient(unauthenticatedMock)
    const { container, debug } = render(
      <ApolloProvider client={client}>
        <App />
      </ApolloProvider>
    )
    await wait()
    expect(container).toMatchSnapshot()
    // TODO: Test incomplete
  })
})
