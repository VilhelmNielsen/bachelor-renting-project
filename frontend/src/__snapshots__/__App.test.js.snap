// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders and matches snapshot 1`] = `
<div>
  <header
    class="css-eqpux9-Header"
  >
    <a
      aria-current="page"
      class="css-prwv7z-Header"
      href="/"
    >
      <img
        alt="Rent it logo"
        class="css-jjhtl6-Header"
        src="logo.svg"
      />
    </a>
    <div
      class="css-1rr4qq7"
    />
    <div
      class="css-nzjyny-Header"
    >
      <div
        class="css-1i1524n-SearchBar"
      >
        <input
          aria-label="search"
          class="css-1dswoqk-SearchBar"
          placeholder="Search..."
          type="text"
          value=""
        />
        
      </div>
      
    </div>
    <div
      class="css-1rr4qq7"
    />
    <nav
      class="css-atca4d-Header"
    >
      <a
        class="css-1h76f4l-MenuItem"
        href="/categories"
      >
        Categories
      </a>
      <a
        class="css-1nosnxj-MenuItem"
        href="/profile"
      >
        Sign in
      </a>
    </nav>
  </header>
  <main
    style="display: none;"
  >
    <div
      role="group"
      style="outline: none;"
      tabindex="-1"
    />
  </main>
  <svg
    height="28"
    stroke="#22a3bb"
    viewBox="0 0 38 38"
    width="28"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      fill="none"
      fill-rule="evenodd"
    >
      <g
        stroke-width="2"
        transform="translate(1 1)"
      >
        <circle
          cx="18"
          cy="18"
          r="18"
          stroke-opacity=".5"
        />
        <path
          d="M36 18c0-9.94-8.06-18-18-18"
        >
          <animatetransform
            attributeName="transform"
            dur="1s"
            from="0 18 18"
            repeatCount="indefinite"
            to="360 18 18"
            type="rotate"
          />
        </path>
      </g>
    </g>
  </svg>
</div>
`;
